File: 3121-count-the-number-of-special-characters-ii/3121-count-the-number-of-special-characters-ii.go
func numberOfSpecialChars(word string) int {
    upperCase := make(map[rune]int)
    lowerCase := make(map[rune]int)
    for i, c := range word {
        if _, ok := upperCase[c]; unicode.IsUpper(c) && !ok{
            upperCase[c] = i
        }

        if unicode.IsLower(c){
            runeChar := []rune(strings.ToUpper(string(c)))
            lowerCase[runeChar[0]] = i
        }
    }

    specials := 0
    for key, uIndex := range upperCase {
        if lIndex, ok := lowerCase[key]; ok && lIndex < uIndex {
            specials++
        }
    }
    return specials
}

File: 3121-count-the-number-of-special-characters-ii/3121-count-the-number-of-special-characters-ii.py
class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        lower_dict = DefaultDict(int)
        upper_dict = DefaultDict(int)
        for i, c in enumerate(word):
            if c.islower():
                lower_dict[c] = i
            
            if c.isupper() and c not in upper_dict:
                upper_dict[c] = i
        
        counts = 0
        for k, v in lower_dict.items():
            if upper_dict[k.upper()] > v:
                counts += 1
        
        return counts
            